compiler/rustc_middle/src/ty/query/plumbing.rs

```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
 
         #[allow(nonstandard_style)]
         #[derive(Clone, Debug)]
-        pub enum Query<$tcx> {
-            $($(#[$attr])* $name($($K)*)),*
+        pub struct Query {
+            pub name: &'static str,
+            hash: Fingerprint,
+            description: Cow<'static, str>,
+            span: Option<Span>,
         }
 
-        impl<$tcx> Query<$tcx> {
-            pub fn name(&self) -> &'static str {
-                match *self {
-                    $(Query::$name(_) => stringify!($name),)*
-                }
-            }
-
-            pub fn describe(&self, tcx: TyCtxt<$tcx>) -> Cow<'static, str> {
-                let (r, name) = match *self {
-                    $(Query::$name(key) => {
-                        (queries::$name::describe(tcx, key), stringify!($name))
-                    })*
+        impl Query {
+            $(#[allow(nonstandard_style)] $(#[$attr])*
+            pub fn $name<$tcx>(tcx: TyCtxt<$tcx>, key: $($K)*) -> Self {
+                let kind = dep_graph::DepKind::$name;
+                let name = stringify!($name);
+                let description = queries::$name::describe(tcx, key);
```

Comment:
It looks like calculating the description is actually relatively expensive, and was previously done on demand but is now done for (each? many?) query invocation - is it worth trying to avoid it if we're not going to need it, perhaps by using an Option here and not computing it in the common case?

cc https://github.com/rust-lang/rust/pull/81901 which switches to using String rather than Cow, which was never borrowed anyway.

Reply:
This function is only called to construct the query backtrace, on exceptional cases: panics and cycles. The number of invocations has not changed, since the number of calls to `describe` used to be exactly the number of instances of the former `Query` enum.